### ISSUE TITLE
WARNING:tensorflow:Input graph does not contain a QueueRunner. That means predict yields forever. This is probably a mistake.

### ISSUE THREAD

**Issue by mari-linhares** [tensorflow/tensorflow#11621]
I have built a Custom Estimator, and I created an input function using the [Datasets API](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/docs_src/programmers_guide/datasets.md) when I call:

``` estimator.predict(input_fn) ```

I see the following Warning:
WARNING:tensorflow:Input graph does not contain a QueueRunner. That means predict yields forever. This is probably a mistake.

The input function looks like this:
```
def get_input_fn(review, word_to_id):
  """Create a Input function for classify_sentiment_analysis.py.

   Args:
      review (str): review sentence
      word_to_id (list): list with all words represented in the embedding.
                         The index is the word index in the embedding.

   Return: sequence of indexes that map that words to the embedding.

  """
  def _word_to_index(sequence):
    """Convert a sequence of words into a sequence of integers."""
    id_sequence = []
    UNK = 399999 # vector for unkown words
    for word in sequence:
      if word in word_to_id:
        id_sequence.append(word_to_id.index(word))
      else:
        id_sequence.append(UNK)
    return np.array(id_sequence)

  def input_fn():
    """Input function."""
    # make review a sequence of words
    review_split = review.split(' ')
    # converting words to indexes
    review_id = _word_to_index(review_split)
    # calculates the length of the sequence
    x_len = len(review_split)
    # creates the dataset from in memory data
    ds = tf.contrib.data.Dataset.from_tensors(review_id)
    # the model expects a batch
    ds = ds.batch(1)

    # creates iterator
    x = ds.make_one_shot_iterator().get_next()

    dict_x = {'x': x, rnn_common.RNNKeys.SEQUENCE_LENGTH_KEY: [x_len]}
    # no label needed since we're only using this input function for prediction
    # if training make sure to return a label
    return dict_x, None

  return input_fn
```

Thank you!

**Comment by shreyneil** [comment]
You are facing this error present in the following link:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/estimator/estimator.py
The condition on line number 491 is satisfied as a result of which you are getting the above error.

Please refer to the blog in order to get the correct idea of an estimator for tensorflow :
http://terrytangyuan.github.io/2016/07/08/understand-and-build-tensorflow-estimator/

Please try the above and get back if any more problem arises.

**Comment by mrry** [comment]
I think your program is perfectly correct and it's the warning that's wrong :). It was added long before the Datasets API existed, and was well intentioned. (AFAICT the `predict()` API relies on your input terminating with a `tf.errors.OutOfRangeError` and before Datasets existed only `QueueRunner`-based pipelines would terminate that way, so it probably *was* an error if no queue runners existed.)

@xiejw Should we remove this warning, or is there a more cunning way to detect the error condition?

**Comment by xiejw** [comment]
I think we should remove this warning. I will do an internal clean up for that then (to and see whether we could this better). 

As mrry@ explained, the error message was added before Datasets API, which served as a nice hint to prevent user from "wrong behavior".

**Comment by mari-linhares** [comment]
Thank you for the quick response 👍

**Comment by contractorwolf** [comment]
wait, how do you solve this issue? I am hitting it like this:
`# Classify two new flower samples.
def new_samples():
  return np.array(
    [[6.4, 3.2, 4.5, 1.5],
     [5.8, 3.1, 5.0, 1.7]], dtype=np.float32)

predictions = list(classifier.predict(input_fn=new_samples))`

**Comment by mari-linhares** [comment]
Hi @contractorwolf,

From where I stand, I think the warning you see "makes sense" in your case, since this implementation suggests a queue input or the datasets API as your input pipeline.

An option would be to use [numpy input function](https://www.tensorflow.org/api_docs/python/tf/estimator/inputs/numpy_input_fn), but not sure if this is still recommended.

**Comment by xiejw** [comment]
mari-linhares  is correct. @contractorwolf's case deserves a warning as the Estimator.predict might not work well with constant arrays. numpy_input_fn should be the best option for now and it will be supported for future releases (see [API stability promises ](https://www.tensorflow.org/programmers_guide/version_compat))

At head, the warning should be removed for Dataset API (already). See code here
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/estimator/estimator.py

I am going to close this issue. Feel free to reopen if it is still a problem.

**Comment by BenjaminLarrousse** [comment]
Hi @xiejw,
I'm facing the same problem as @mari-linhares with my custom estimator (and an input function build with the tf.dataset API).
I see that the warning is removed in the master branch of Tensorflow. Any ideas where it would be in a release ? 
In the mean time, is there any work around I can use ?
Thanks

**Comment by contractorwolf** [comment]
I figured I should leave a sample that worked for me, after I read what @xiejw wrote (thanks for the advice).  This solved it for me with the Iris dataset:

# test using single records (2 real world values)
# one Iris-setosa, one Iris-versicolor
records_to_predict = pd.DataFrame([
    [5.1, 3.5, 1.4, 0.2], # first records features
    [5.7, 2.8, 4.1, 1.3]],# second record features
    columns=columns_names)# defined column names
 
# predict using the model to get a prediction for each item passed
saved_prediction_input_fn = tf.estimator.inputs.pandas_input_fn(records_to_predict, num_epochs=1, shuffle=False)

# make predictions and store as a list
saved_prediction_result = list(saved_estimator.predict(input_fn=saved_prediction_input_fn))

**Comment by lorenzoriano** [comment]
It should be noted that in the [Estimator documentation](https://www.tensorflow.org/api_docs/python/tf/estimator/Estimator#predict) the `input_fn` could return a `tf.Tensor` (or a tuple). I initially wrote my code so that the `input_fn` would return just that, but `predict` generates an infinite iterator.

Am I misinterpreting the docs or do they need update?

cc @xiejw @mrry

**Comment by bobbyAtSperry** [comment]
Yes, I have the same problem as @lorenzoriano . I am returning a tf.data.Dataset.from_tensor_slices and get the error message at the top of this thread. Any suggestions on how I can fix this without rewriting to not use tf.Tensor which would feel like a backward step.

**Comment by agniszczotka** [comment]
I experinced the same warring when using :
`dataset = tf.data.Dataset().from_generator(_get_generator, output_types=(tf.float32),
                                                   output_shapes=(tf.TensorShape([None, None, None, None, 1])),
                                                   ).repeat(1)`

`iterator = dataset.make_one_shot_iterator()`

`t = iterator.get_next()`
